### PR TITLE
feat: Add Github Action for notation setup

### PR DESCRIPTION
**Pull request by duffney**
Contributing [setup-notation](https://github.com/duffney/setup-notation/tree/main) as notation-action.

### CODE REVIEW

File: setup/action.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    default: self-signed
+  certificate_key_id:
+    description: Key identifier for the signing certificate
+    required: false
+  plugin_name:
+    description: Name of the Notation plugin to install
+    required: false
+  plugin_version:
+    description: Version of the Notation plugin to install
+    required: false
```
Comment (Two-Hearts) [inline setup/action.yml:23]: These inputs are related to `sign` and/or `verify`, I think we should remove them from the `setup` action. 
In addition, we should add a new input called `url` for installing notation CLI via a URL. For example, if I want to build a release of Notation CLI on my local branch and use it in the rest of the workflow, the current action cannot help me to do that.
```suggestion
inputs:
  version:
    description: Version of Notation CLI to install
    required: false
    default: 1.0.0-rc.7
  url:
    description: URL of Notation CLI to install
    required: false
```

File: setup/index.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
+const path = require('path');
+const core = require('@actions/core');
+const tc = require('@actions/tool-cache');
+const { getDownloadObject, getDownloadPluginObject, addPluginCert, versionCheck } = require('./lib/utils');
+const { get } = require('http');
```
Comment (Two-Hearts) [inline setup/index.js:5]: Looks like we are not using this one?
```suggestion
```

File: setup/index.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+const fs = require('fs');
+const mv = require('mv');
+const { version } = require('os');
+const execSync = require('child_process').execSync;
+
+async function setup() {
+  try {
+    // Get version of tool to be installed
+    const version = core.getInput('version'); //"0.9.0-alpha.1";
```
Comment (Two-Hearts) [inline setup/index.js:14]:
```suggestion
    const version = core.getInput('version');
```

File: setup/index.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    const keyName = core.getInput('key_name');
+    
+    // Check if the NotationCli version is supported
+    versionCheck(version);
+
+    // Check if the NotationCli version is supported
+    versionCheck(version);
```
Comment (Two-Hearts) [inline setup/index.js:21]: Do we need to check the version twice?
```suggestion
    // Check if the NotationCli version is supported
    versionCheck(version);
```

File: setup/package.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  "name": "setup-notation",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/Duffney/setup-notation.git"
```
Comment (JeyJeyGao) [inline setup/package.json:11]: Do we need to update the url for the new repository?

File: setup/index.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    const pathToTarball = await tc.downloadTool(download.url);
+
+    // Extract the tarball/zipball onto host runner
+    const extract = download.url.endsWith('.zip') ? tc.extractZip : tc.extractTar;
+    const pathToCLI = await extract(pathToTarball);
+    fs.mkdirSync(pathToCLI + "/" + download.binPath, { recursive: true, })
+
+    const currentPath = path.join(pathToCLI, "notation")
+    const destinationPath = path.join(pathToCLI, "/", download.binPath, "/", "notation")
+    fs.rename(currentPath, destinationPath, function (err) {
```
Comment (Two-Hearts) [inline setup/index.js:35]: Should we use `mv` here?
```suggestion
    mv(currentPath, destinationPath, function (err) {
```

File: setup/index.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    const pluginName = core.getInput('plugin_name');
+    const pluginVersion = core.getInput('plugin_version');
+
+    if (pluginName) {
+      setupPlugin(pluginName, pluginVersion, keyName);
+    } else {
+      // Generate a local signing certificate
+      const output = execSync(`notation cert generate-test --default "${keyName}"`, { encoding: 'utf-8' });
+      console.log('notation cert output:\n', output);
+    }
```
Comment (Two-Hearts) [inline setup/index.js:55]: As we discussed, we should keep plugin related logic in `sign` and `verify`. This is because `sign` and `verify` can have very different configurations in a single workflow. `notation-setup` should focus on setting up Notation CLI. 
(Note: Let's remove the `notation cert generate-test` from the action. It is designed to be used in tests ONLY as it generates a self-signed certificate, which cannot be used in production.)
```suggestion 
```

File: setup/lib/utils.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+const path = require('path');
+const execSync = require('child_process').execSync;
+const semver = require('semver');
+const core = require('@actions/core');
+
+// arch in [arm, x32, x64...] (https://nodejs.org/api/os.html#os_os_arch)
+// return value in [amd64, 386, arm]
+function mapArch(arch) {
+    const mappings = {
+        x32: '386',
```
Comment (JeyJeyGao) [inline setup/lib/utils.js:12]: Looks like Notation doesn't have `x32` binary according to the [release page](https://github.com/notaryproject/notation/releases). We only have `amd64` and `arm64`

File: setup/lib/utils.js [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    };
+    return mappings[os] || os;
+}
+
+function getDownloadObject(version) {
+    const platform = os.platform();
+    const filename = `notation_${version}_${mapOS(platform)}_${mapArch(os.arch())}`;
+    const extension = platform === 'win32' ? 'zip' : 'tar.gz';
+    const binPath = platform === 'win32' ? 'bin' : path.join(filename, 'bin');
+    const url = `https://github.com/notaryproject/notation/releases/download/v${version}/${filename}.${extension}`;
```
Comment (Two-Hearts) [inline setup/lib/utils.js:33]: As mentioned above, making `url` as an input is more flexible.

File: setup/lib/utils.js [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        x64: 'amd64'
+    };
+    return mappings[arch] || arch;
+}
+
+// os in [darwin, linux, win32...] (https://nodejs.org/api/os.html#os_os_platform)
+// return value in [darwin, linux, windows]
+function mapOS(os) {
+    const mappings = {
+        darwin: 'macOS',
```
Comment (JeyJeyGao) [inline setup/lib/utils.js:22]: Looks like we don't need to map the name for `darwin`: `notation_1.0.0-rc.7_darwin_amd64.tar.gz`

File: setup/lib/utils.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+const core = require('@actions/core');
+
+// arch in [arm, x32, x64...] (https://nodejs.org/api/os.html#os_os_arch)
+// return value in [amd64, 386, arm]
+function mapArch(arch) {
+    const mappings = {
+        x32: '386',
+        x64: 'amd64'
+    };
+    return mappings[arch] || arch;
```
Comment (JeyJeyGao) [inline setup/lib/utils.js:15]: We may need an exception for unknown `arch`

File: setup/lib/utils.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+// os in [darwin, linux, win32...] (https://nodejs.org/api/os.html#os_os_platform)
+// return value in [darwin, linux, windows]
+function mapOS(os) {
+    const mappings = {
+        darwin: 'macOS',
+        win32: 'windows'
+    };
+    return mappings[os] || os;
```
Comment (JeyJeyGao) [inline setup/lib/utils.js:25]: We may need an exception for unknown `os`

File: setup/lib/utils.js [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+function addPluginCert(keyName,keyId){
+    if (keyId.includes('vault.azure.net')){
+        const output = execSync(`notation key add --plugin azure-kv --id ${keyId} --default ${keyName}`, { encoding: 'utf-8' });
+        console.log('notation cert output:\n', output);
+    } // Add logic for additional Notation plugins here
+}
+
+function versionCheck(version){
+    const supportedVersion = '0.12.0-beta.1'
```
Comment (Two-Hearts) [inline setup/lib/utils.js:67]: Let's update the supported versions: https://github.com/notaryproject/notation/releases

File: setup/package.json [hunk 14]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "setup-notation",
```
Comment (Two-Hearts) [inline setup/package.json:2]: how about
```suggestion
  "name": "notation-setup",
```

File: setup/lib/utils.js [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+    const mappings = {
+        x32: '386',
+        x64: 'amd64'
+    };
+    return mappings[arch] || arch;
+}
+
+// os in [darwin, linux, win32...] (https://nodejs.org/api/os.html#os_os_platform)
+// return value in [darwin, linux, windows]
+function mapOS(os) {
```
Comment (JeyJeyGao) [inline setup/lib/utils.js:20]: Can we change the name of the variable `os` because it is a built-in library name.

File: setup/package.json (hunk 5 above)
Comment (Two-Hearts) [inline setup/package.json:11]: Shall we change the url to this Github repo?

File: setup/package.json [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+    "type": "git",
+    "url": "git+https://github.com/Duffney/setup-notation.git"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
+  "bugs": {
+    "url": "https://github.com/Duffney/setup-notation/issues"
+  },
+  "homepage": "https://github.com/Duffney/setup-notation#readme",
```
Comment (Two-Hearts) [inline setup/package.json:19]: Same here.

File: setup/package.json [hunk 17]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "setup-notation",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
```
Comment (shizhMSFT) [inline setup/package.json:5]: It is better to upgrade to TypeScirpt. We can do that in another PR.

File: setup/package.json [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
```
Comment (shizhMSFT) [inline setup/package.json:7]: It is better to have a `build` script for quicker build just like a `Makefile`.

File: setup/package.json [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/Duffney/setup-notation.git"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
```
Comment (shizhMSFT) [inline setup/package.json:15]: Should it be Apache 2.0?

File: setup/action.yml (hunk 1 above)
Comment (shizhMSFT) [reply setup/action.yml:23]: +1

File: setup/lib/utils.js [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+    } // Add logic for additional Notation plugins here
+}
+
+function versionCheck(version){
+    const supportedVersion = '0.12.0-beta.1'
+    if (semver.lt(version, supportedVersion)) {
+        core.setFailed('Unsupported NotationCli version');
+        throw new Error(`NotationCli v${version} is not supported by this version of the setup_notation GitHub Action.`);
+    }
+}
```
Comment (shizhMSFT) [inline setup/lib/utils.js:72]: Instead of semver comparation, we can maintain an allow list of supported version with checksums.

File: setup/index.js [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+  try {
+    const keyId = core.getInput('certificate_key_id');
+    const download = getDownloadPluginObject(name, version)
+    console.log(download)
+
+    fs.mkdirSync(download.pluginPath, { recursive: true, })
+
+    const pathToTarball = await tc.downloadTool(download.url);
+    const extract = download.url.endsWith('.zip') ? tc.extractZip : tc.extractTar;
+    const pathToPluginDownload = await extract(pathToTarball);
```
Comment (shizhMSFT) [inline setup/index.js:72]: The integrity is required to be verified against a checksum after downloading before extraction.

File: setup/package.json (hunk 18 above)
Comment (Two-Hearts) [inline setup/package.json:7]: Here's an example of a build script:
`ncc build index.ts -o dist --source-map --minify --license licenses.txt`

File: setup/lib/utils.js (hunk 8 above)
Comment (duffney) [reply setup/lib/utils.js]: Add an arch const with the value of amd64 and removed the mapArch function. :)

File: setup/lib/utils.js (hunk 10 above)
Comment (duffney) [reply setup/lib/utils.js:13]: Is this because the binary name uses the term `darwin`, instead of `macOS`? And because the `os.platform()` call returns `darwin` there's no need to map the values?

File: setup/lib/utils.js (hunk 11 above)
Comment (duffney) [reply setup/lib/utils.js]: Are you thinking of something like this?

```
function mapOS(os) {
    const mappings = {
        darwin: 'macOS',
        win32: 'windows'
    };

    if (os in mappings) {
        return mappings[os];
    } else {
        return 'Unknown OS';
    }
}
```

File: setup/lib/utils.js (hunk 12 above)
Comment (duffney) [reply setup/lib/utils.js:16]: Are you thinking of something like this?

```
function mapOS(os) {
    const mappings = {
        darwin: 'macOS',
        win32: 'windows'
    };

    if (os in mappings) {
        return mappings[os];
    } else {
        return 'Unknown OS';
    }
}
```

File: setup/lib/utils.js (hunk 11 above)
Comment (duffney) [reply setup/lib/utils.js]: I remove this func with a const with the value of `arm64`, but it might be better to replace it with a call to `os.Arch` to detect if it's amd or arm and append that to the `filename`.

File: setup/index.js (hunk 4 above)
Comment (duffney) [reply setup/index.js]: Nope, that was a typo. :) Removed it.

File: setup/package.json (hunk 16 above)
Comment (duffney) [reply setup/package.json]: Remove this line because I didn't see it in the other actions like sign and verify.

File: README.md
[the remark this answers is not in the document]
Comment (duffney) [reply README.md:1]: I added the README.md from the root of the repo to avoid it being deleted. Thanks for calling this out.

File: setup/lib/utils.js (hunk 20 above)
Comment (duffney) [reply setup/lib/utils.js:63]: Would it be best for me to remove the `versionCheck` function and add the checksum with an additional PR later?

File: setup/lib/utils.js (hunk 13 above)
Comment (duffney) [reply setup/lib/utils.js:58]: Further down in the thread @shizhMSFT mentioned replacing this with a checksum. What do you think about removing this and adding a checksum in a sperate PR later?

File: setup/package.json (hunk 5 above)
Comment (duffney) [reply setup/package.json]: Done. :)

File: setup/package.json (hunk 17 above)
Comment (duffney) [reply setup/package.json:5]: Should I open an issue for this on the repo?

File: setup/index.js (hunk 21 above)
Comment (duffney) [reply setup/index.js:68]: @Two-Hearts mentioned merging these action into the repo and then doing a refactoring on them with better separation of concerns. Would it be best to wait to implement the checksum until the refactoring?

File: setup/package.json (hunk 18 above)
Comment (duffney) [reply setup/package.json:7]: Should that be added into the scripts section within the value of the test property?

File: setup/action.yml (hunk 1 above)
Comment (Two-Hearts) [reply setup/action.yml:23]: > To clarity, all plugin setup is now going to be done by the sign action?

@duffney Yes, `sign` and `verify` would have their own steps of configuring plugins. `setup` will only set up the Notation CLI.

File: setup/lib/utils.js (hunk 13 above)
Comment (Two-Hearts) [reply setup/lib/utils.js:58]: > Further down in the thread @shizhMSFT mentioned replacing this with a checksum. What do you think about removing this and adding a checksum in a sperate PR later?

Yeah, we should check the SHA256 checksum after the download and before extracting. We could have a separate PR for it. And I think we should keep the version check? When the user input is just a Notation CLI version, then we should check the version before trying to download it to save the roundtrip.

File: setup/package.json (hunk 18 above)
Comment (Two-Hearts) [reply setup/package.json:7]: > Should that be added into the scripts section within the value of the test property?

Oh, it's not for test. Should be something like this:
```suggestion
  "scripts": {
      "build": "ncc build index.ts -o dist --source-map --minify --license licenses.txt",
      "test": "echo \"Error: no test specified\" && exit 1"
   }
```

File: setup/index.js (hunk 21 above)
Comment (Two-Hearts) [reply setup/index.js:68]: > @Two-Hearts mentioned merging these actions into the repo and then doing a refactoring on them with better separation of concerns. Would it be best to wait to implement the checksum until the refactoring?

Sure. Actually, this logic has already been added in the `sign` PR. I will create a separate PR to include duplicated logics in one place (after current PRs got merged).

File: setup/lib/utils.js (hunk 20 above)
Comment (Two-Hearts) [reply setup/lib/utils.js:63]: @shizhMSFT I'm thinking if we should keep the version check? When the user input is just a Notation CLI version, then we could check the version before any download to save the roundtrip. After a valid download and before extracting, we could do the checksum check. What do you think?

File: setup/lib/utils.js (hunk 13 above)
Comment (shizhMSFT) [reply setup/lib/utils.js:58]: We need to maintain a `version`-to-`checksum` map. If a version is not in that map, it is "unsupported" and thus a checksum should be provided by the user. We can do that in another PR.

File: setup/lib/utils.js (hunk 20 above)
Comment (shizhMSFT) [reply setup/lib/utils.js:63]: The `version`-to-`checksum` map itself is a version check

Comment (shizhMSFT) [reply setup/lib/utils.js:63]: > Would it be best for me to remove the `versionCheck` function and add the checksum with an additional PR later?

SGTM

File: setup/lib/utils.js (hunk 13 above)
Comment (Two-Hearts) [reply setup/lib/utils.js:58]: > We need to maintain a `version`-to-`checksum` map. If a version is not in that map, it is "unsupported" and thus a checksum should be provided by the user. We can do that in another PR.

Oh yeah, good idea!

File: setup/lib/utils.js (hunk 15 above)
Comment (duffney) [reply setup/lib/utils.js:11]: @Two-Hearts, would this be a good item to add to the refactor list?

File: setup/action.yml (hunk 1 above)
Comment (duffney) [reply setup/action.yml:23]: I'm uncertain how much value the `url` input will add and how much it will be used. Even from the perspective of a developer working on Notation, I think the chances of them needing to use a dev branch within a workflow are low and the odds of a consumer of the action needing it to be even lower. With that in mind does it add enough value to add in? I could have the wrong perspective and be totally wrong, so please feel free to correct my understanding.

File: setup/package.json (hunk 18 above)
Comment (duffney) [reply setup/package.json:7]: Oh! that makes life so much easier! Clearly, my lack of JS knowledge is showing. :) Thank you @Two-Hearts

File: setup/lib/utils.js (hunk 13 above)
Comment (duffney) [reply setup/lib/utils.js]: I removed the `versionCheck` code. :)

File: setup/lib/utils.js (hunk 20 above)
Comment (duffney) [reply setup/lib/utils.js]: versionCheck removed and a checksum will be add in a future PR after the actions are merged.

File: setup/lib/utils.js (hunk 15 above)
Comment (Two-Hearts) [reply setup/lib/utils.js:11]: > @Two-Hearts, would this be a good item to add to the refactor list?

@duffney Sure thing, because other PRs may need the same change.

File: setup/package.json (hunk 18 above)
Comment (Two-Hearts) [reply setup/package.json]: > Oh! that makes life so much easier! Clearly, my lack of JS knowledge is showing. :) Thank you @Two-Hearts

No problem @duffney. I'm also in the learning phase :)

File: setup/package.json [hunk 22]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "notation-setup",
+  "version": "1.0.0",
+  "description": "Install and configure the Notation CLI and plugins.",
+  "main": "index.js",
+  "scripts": {
+  "scripts": {
+      "build": "ncc build index.js -o dist --source-map --minify --license licenses.txt",
```
Comment (Two-Hearts) [inline setup/package.json:8]: @duffney I see you updated the package.json, could you run the `npm run build` under `setup` dir?

File: setup/package.json (hunk 18 above)
Comment (duffney) [reply setup/package.json:8]: build has been added to the package.json

File: setup/lib/utils.js (hunk 15 above)
Comment (duffney) [reply setup/lib/utils.js:11]: @Two-Hearts should I open this as an issue on the repo or do you have a running list for the refactor work?

File: setup/lib/utils.js (hunk 12 above)
Comment (duffney) [reply setup/lib/utils.js:16]: This might also be a good candidate for the refactor work once the PR is merged.

File: setup/package.json (hunk 22 above)
Comment (duffney) [reply setup/package.json]: Ran `npm run build` and pushed up changes. :) Thank you for the instructions @Two-Hearts.

File: setup/index.js (hunk 6 above)
Comment (Two-Hearts) [reply setup/index.js:28]: @duffney, any thoughts on this change?

File: setup/lib/utils.js (hunk 15 above)
Comment (Two-Hearts) [reply setup/lib/utils.js:11]: > @Two-Hearts should I open this as an issue on the repo or do you have a running list for the refactor work?

@duffney yeah, please open an issue on it. Thanks.

File: setup/index.js [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+      const output = execSync(`notation cert generate-test --default "${keyName}"`, { encoding: 'utf-8' });
+      console.log('notation cert output:\n', output);
+    }
+
+  } catch (e) {
+    core.setFailed(e);
+  }
+}
+
+async function setupPlugin(name, version, keyName) {
```
Comment (Two-Hearts) [inline setup/index.js:55]: We should also remove this function from `setup`.

File: setup/index.js [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+        addPluginCert(keyName,keyId);
+      }
+    });
+  } catch (e) {
+    core.setFailed(e);
+  }
+}
+
+module.exports = setup
+module.exports = setupPlugin
```
Comment (Two-Hearts) [inline setup/index.js:88]: same here.

File: setup/lib/utils.js [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+            url,
+            filename,
+            pluginPath
+        };
+    } else { // Add logic for additional Notation plugins here
+        setFailed(`Plugin ${name} is not currently supported`)
+    }
+}
+
+function addPluginCert(keyName,keyId){
```
Comment (Two-Hearts) [inline setup/lib/utils.js:50]: Same here.

File: setup/lib/utils.js [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+    const binPath = platform === 'win32' ? 'bin' : path.join(filename, 'bin');
+    const url = `https://github.com/notaryproject/notation/releases/download/v${version}/${filename}.${extension}`;
+    return {
+        url,
+        binPath,
+        filename
+    };
+}
+
+function getDownloadPluginObject(name, version) {
```
Comment (Two-Hearts) [inline setup/lib/utils.js:32]: Same here.

File: setup/package.json (hunk 17 above)
Comment (FeynmanZhou) [reply setup/package.json:5]: We can create an issue in this repo and improve it in a follow-up PR

File: setup/lib/utils.js (hunk 15 above)
Comment (duffney) [reply setup/lib/utils.js:11]: Issue opened with a link to this thread. 

https://github.com/notaryproject/notation-action/issues/6

File: setup/index.js (hunk 6 above)
Comment (duffney) [reply setup/index.js:28]: LGTM

File: setup/lib/utils.js (hunk 9 above)
Comment (duffney) [reply setup/lib/utils.js:24]: Is the suggestion to replace `https://github.com/notaryproject/notation/releases/download/v` with the input from the `url` parameter from the action?

So the const would look like this:

```js
const downloadUrl = core.getInput('url');
   const url = `${downloadUrl}/${version}/${filename}.${extension}`;
```

And then the actions.yml could be updated to have the default of:

```yml
  url:
    description: URL of Notation CLI to install
    required: false
    default: https://github.com/notaryproject/notation/releases/download/v
```

Comment (Two-Hearts) [reply setup/lib/utils.js:24]: For user input `url`, we could make it a full valid url (without default) and add an input called `checksum` in `action.yml`. If user inputs a `url`, `version` will be skipped. The github action downloads Notation CLI from the url directly and check the checksum based on user input (we could add this logic in the refactoring PR later, this logic will be duplicated in many places). So something like this:
```
 url:
    description: URL of Notation CLI to install
    required: false
 checksum:
    description: checksum of the downloaded Notation CLI tarball. Required if `url` is provided.
    required: false
```

File: setup/lib/utils.js [hunk 27]
```diff
@@ -0,0 +1,32 @@
+const { setFailed } = require('@actions/core');
+const os = require('os');
+const path = require('path');
+const execSync = require('child_process').execSync;
+const semver = require('semver');
```
Comment (Two-Hearts) [inline setup/lib/utils.js:5]: Looks like these two are not used. We can remove them.

File: setup/index.js [hunk 28]
```diff
@@ -0,0 +1,47 @@
+const path = require('path');
+const core = require('@actions/core');
+const tc = require('@actions/tool-cache');
+const { getDownloadObject } = require('./lib/utils');
+const fs = require('fs');
+const mv = require('mv');
+const { version } = require('os');
+const execSync = require('child_process').execSync;
```
Comment (Two-Hearts) [inline setup/index.js:8]: Looks like these two are not used. We can remove them.

Comment (duffney) [reply setup/index.js]: Removed with the latest commit, thanks for catching that. :)

File: setup/lib/utils.js (hunk 27 above)
Comment (duffney) [reply setup/lib/utils.js]: Removed with the latest commit, thanks for catching that. :)

File: setup/lib/utils.js (hunk 10 above)
Comment (JeyJeyGao) [reply setup/lib/utils.js:9]: Yes, you can find all the artifacts' name here: https://github.com/notaryproject/notation/releases/tag/v1.0.0-rc.7

File: setup/lib/utils.js (hunk 12 above)
Comment (JeyJeyGao) [reply setup/lib/utils.js:12]: Yes, similar to your example code. We also need to consider the case for `linux`

```typescript
function getPlatform(): string {
    const platform: string = os.platform();
    switch (platform) {
        case 'linux':
            return 'linux';
        case 'darwin':
            return 'darwin';
        case 'win32':
            return 'windows';
        default:
            throw new Error(`Unsupported platform: ${platform}`);
    }
}
```

File: setup/lib/utils.js (hunk 9 above)
Comment (Two-Hearts) [reply setup/lib/utils.js:20]: > Imo, with the addition of the checksum input this work should be done under the PR for the checksum implementation with the refactor work. If I add the checksum input now there won't be any logic behind it with the current version of the code base.
> 
> Completely agree with the change though. :)

Sure. I agree with you @duffney. Please feel free to resolve the comment (somehow I'm not able to resolve it from my side).

**Review by Two-Hearts — APPROVED** [review]
LGTM

File: setup/lib/utils.js [hunk 29]
```diff
@@ -0,0 +1,28 @@
+const os = require('os');
+const path = require('path');
+const arch = "amd64";
```
Comment (shizhMSFT) [inline setup/lib/utils.js:3]: There can be other architectures if self-hosted runners are used.

Ref: https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/about-self-hosted-runners#architectures

**Review by shizhMSFT — APPROVED** [review]
LGTM as the first iteration. We should convert all unresolved conversation as issues, merge this PR, and open other PRs to address them.

**Comment by Two-Hearts** [comment]
@duffney There's a conflict in the README file of your PR. Could you resolve it so that we can merge? Thanks.

File: setup/action.yml [hunk 30]
```diff
@@ -0,0 +1,26 @@
+name: setup-gh
```
[the remark this answers is not in the document]
Comment (Two-Hearts) [reply setup/action.yml:1]: Issue created.

File: setup/lib/utils.js (hunk 9 above)
Comment (Two-Hearts) [reply setup/lib/utils.js:20]: Issue created: https://github.com/notaryproject/notation-action/issues/11

File: setup/package.json (hunk 17 above)
Comment (Two-Hearts) [reply setup/package.json:5]: Issue created: https://github.com/notaryproject/notation-action/issues/12

File: setup/lib/utils.js (hunk 29 above)
Comment (Two-Hearts) [reply setup/lib/utils.js:3]: Issue created: https://github.com/notaryproject/notation-action/issues/13

File: setup/lib/utils.js (hunk 12 above)
Comment (Two-Hearts) [reply setup/lib/utils.js:12]: Issue created: https://github.com/notaryproject/notation-action/issues/13
